### PR TITLE
Allow user to set default reminder duration for new events

### PR DESCRIPTION
**Pull request by tchernobog**
This PR implements a new setting in the calendar web UI to allow the user to specify a default reminder to apply to newly created events.

This is particularly helpful when you want to create a batch of events in a row.

It should provide a workable solution to issue #629, albeit it is a global setting rather than per-calendar.

### CODE REVIEW

**Comment by tchernobog** [comment]
Hi @ChristophWurst, do you know who could perform a review? Thanks!

**Review by tcitworld — COMMENTED** [review]
I'm conflicted about this. It's definitely not enough to close #629 since it's only a setting to add a simple timed based `DISPLAY` reminder and it doesn't consider events to be `allDay` to suggest appropriate slots. Also it crowds even more the bottom left settings area since we don't have #2304 yet.

Still, might be handy for some people in the mean time, so :shrug:

File: lib/Controller/SettingsController.php [hunk 1]
```diff
@@ -310,4 +312,29 @@ private function setSlotDuration(string $value):JSONResponse {
...
 	}
+
+	/**
+	 * sets defaultReminder for user
+	 *
+	 * @param string $value User-selected option for default_reminder in agenda view
+	 * @return JSONResponse
+	 */
+	private function setDefaultReminder(string $value):JSONResponse {
+		if ($value !== 'none' && !(is_numeric($value) && $value <= 0)) {
```
[the remark this answers is not in the document]
Comment (tchernobog) [reply lib/Controller/SettingsController.php:323]: `is_int` however does not convert from a string to a number like `is_numeric` does (and the input is always a string).

**Comment by tchernobog** [comment]
> I'm conflicted about this. It's definitely not enough to close #629 since it's only a setting to add a simple timed based `DISPLAY` reminder and it doesn't consider events to be `allDay` to suggest appropriate slots. Also it crowds even more the bottom left settings area since we don't have #2304 yet.
> 
> Still, might be handy for some people in the mean time, so shrug

I know, and maybe I should reword the commit to avoid saying it fixes #629. But it was enough to cover my personal needs of having a default reminder set when creating events through the Web UI, which is the specific improvement I sought. I am not going after the bug bounty anyway.

File: lib/Controller/SettingsController.php (hunk 1 above)
[the remark this answers is not in the document]
Comment (tcitworld) [reply lib/Controller/SettingsController.php:323]: `filter_var($value, FILTER_VALIDATE_INT) !== false` could be a better fit then.

[the remark this answers is not in the document]
Comment (tchernobog) [reply lib/Controller/SettingsController.php]: Added `filter_var` in latest version; also removed from commit message comment that it fixes #629.

File: lib/Controller/SettingsController.php [hunk 2]
```diff
@@ -310,4 +312,31 @@ private function setSlotDuration(string $value):JSONResponse {
...
+		}
+
+		try {
+			$this->config->setUserValue(
+				$this->userId,
+				$this->appName,
+				'defaultReminder',
+				$value
+			);
+		} catch (\Exception $e) {
```
Comment (ChristophWurst) [inline lib/Controller/SettingsController.php:336]: this is not an expected exception, right?

Comment (tchernobog) [reply lib/Controller/SettingsController.php:336]: No, it just mirrors error handling as with all the other `set...` functions (e.g. `setSlotDuration`).

Looking at the manual, I see other examples from https://docs.nextcloud.com/server/latest/developer_manual/basics/storage/configuration.html don't do error handling, and https://github.com/nextcloud/server/blob/master/lib/public/IConfig.php doesn't mention any thrown exception from setUserValue(). 

I don't know if every similar try/catch should be dropped.

Comment (ChristophWurst) [reply lib/Controller/SettingsController.php:336]: ok, then let's keep it for consistency

**Comment by tchernobog** [comment]
Hi @ChristophWurst, are there more changes you'd like me to perform? Thanks!

**Comment by tchernobog** [comment]
@ChristophWurst, @tcitworld: any further feedback you'd like me to address? Thanks!

**Comment by tchernobog** [comment]
Weekly ping :-)

**Comment by tchernobog** [comment]
Hi @ChristophWurst, any major blocker to see this merged? I wanted to contributed a few more enhancements to the calendar app, but I would first like to see how it goes with this PR.

**Review by ChristophWurst — APPROVED** [review]
Code looks good!

**Comment by szaimen** [comment]
🚀

**Pull request merged**